### PR TITLE
Build Active Directory interaction library separately

### PR DESCRIPTION
**Pull request by NIR-Ginko**
This pull request moves code for interaction with Active Directory LDAP database into a separate library in `adldap` directory. Here we're building `libadldap_c.so` and `libadldap_cpp.so`. The difference is API and linkage with `libstdc++`. This gives us freedom to use the same code from both C and C++.

Please note that code is ready for peer review but cleanup and C++ wrapper is still Work In Progress.

### CODE REVIEW

File: adtool.spec [hunk 1]
```diff
@@ -52,6 +52,8 @@ cd BUILD
 %files
 %doc README.md
 %_bindir/adtool
+%_libdir/libadldap_c.so
```
Comment (mastersin) [inline adtool.spec:55]: А давайте без _c и _cpp тогда
libadldap.so
libadldap++.so

File: config.h.in [hunk 2]
```diff
@@ -36,6 +36,7 @@
 #   define ADTOOL_ORGANIZATION_DOMAIN "${ADTOOL_ORGANIZATION_DOMAIN}"
 
 #   define HEAD_DN "DC=domain,DC=alt"
+#   define SEARCH_BASE "ldap://dc0.domain.alt"
```
Comment (mastersin) [inline config.h.in:39]: По моему, SEARCH_BASE - это тоже, что и HEAD_DN:
```
$ ldapsearch --help 2>&1 | grep '[hH].*\s*LDAP'
  -h host    LDAP server
  -H URI     LDAP Uniform Resource Identifier(s)
```

File: adldap/include/ad_connection.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+#define __ADLDAP_AD_CONNECTION_H 1
+
+#include "active_directory.h"
+
+#include <string>
+
+namespace adldap {
+
+class AdConnection {
+    std::string uri;
```
Comment (Kvel2D) [inline adldap/include/ad_connection.h:30]: Why use std::string here? This layer is application only, so should use QString.

**Comment by Kvel2D** [comment]
AdConnection being separate from AdInterface seems wrong - to call AdInterface functions you need a connection. Should move AdInterface functions into AdConnection and pass uri directly to active_directory.h functions, no boilerplate of functions in AdConnection this way as well.
Also pass ldap connection(ds) instead of uri? That's what active_directory functions ultimately need, they don't use the uri string itself for anything.

Also

File: adtool.spec (hunk 1 above)
Comment (NIR-Ginko) [reply adtool.spec]: Done.

File: config.h.in (hunk 2 above)
Comment (NIR-Ginko) [reply config.h.in:39]: Idunno. Let's leave this part for the next iteration with C API redesign.

File: adldap/include/ad_connection.h (hunk 3 above)
Comment (NIR-Ginko) [reply adldap/include/ad_connection.h:30]: This is Qt-independent code available for linking with any application that needs AD LDAP connection.

Comment (Kvel2D) [reply adldap/include/ad_connection.h:30]: Ok.

File: config.h.in (hunk 2 above)
Comment (Kvel2D) [reply config.h.in:39]: HEAD_DN shouldn't really be a constant, I used it as constant for convenience. Will figure out how to get it with LDAP requests.

**Pull request merged**